src/prove.rs

```diff
@@ -147,21 +147,28 @@ impl Prover8_56 {
             "nonces must be a multiple of 16"
         );
         log::info!("calculating proof of work for nonces {nonces:?}",);
-        let ciphers: Vec<AesCipher> = nonce_group_range(nonces.clone(), Self::NONCES_PER_AES)
-            .map(|nonce_group| {
-                log::debug!("calculating proof of work for nonce group {nonce_group}");
-                let pow = pow_prover.prove(
-                    nonce_group.try_into()?,
-                    challenge[..8].try_into().unwrap(),
-                    &params.pow_difficulty,
-                    miner_id,
-                )?;
-                log::debug!("proof of work: {pow}");
-
-                Ok(AesCipher::new(challenge, nonce_group, pow))
-            })
-            .collect::<eyre::Result<_>>()?;
+        let map_fn = |nonce_group: u32| -> eyre::Result<AesCipher> {
+            log::debug!("calculating proof of work for nonce group {nonce_group}");
+            let pow = pow_prover.prove(
+                nonce_group.try_into()?,
+                challenge[..8].try_into().unwrap(),
+                &params.pow_difficulty,
+                miner_id,
+            )?;
+            log::debug!("proof of work: {pow}");
+
+            Ok(AesCipher::new(challenge, nonce_group, pow))
+        };
 
+        let ciphers: Vec<AesCipher> = match pow_prover.par() {
+            true => nonce_group_range(nonces.clone(), Self::NONCES_PER_AES)
+                .into_par_iter()
+                .map(map_fn)
+                .collect::<eyre::Result<_>>()?,
+            false => nonce_group_range(nonces.clone(), Self::NONCES_PER_AES)
+                .map(map_fn)
+                .collect::<eyre::Result<_>>()?,
+        };
```

Comment:
Now that I think of it again, using rayon has a downside of limiting parallel execution to the parallelism available on the **local** machine, while it should be limited by the number of available backend services. I.e. if the local machine has 2 cores it will run 2 k2pows in parallel, even if there are more k2pow services available. On the other hand, we can't know how many k2pow services there are. Perhaps it should just spawn a thread for every nonce_group that needs a k2pow? IDK if rayon supports "unlimited" parallelism.

Reply:
Yeah I had some doubts about how this would work. The problem is a bit more nuanced, and I think that `rayon` and the parallel iterator are misused here, since they aim to solve a different problem here.
The thing is that we don't need a thread pool here. Work stealing with little threads would do the same thing with far less context switching. `async` would be superior here, because:
- the request to the API that provides the response involves almost entirely of I/O on the network side.
- having multiple blocking threads on this sort of task is extremely wasteful.
- it provides back-pressure indeed, but in a very ugly way and with a huge overhead of the thread-pool.

Speaking purely on what we probably want in this specific case is the following:
- have a regular iterator that spawns multiple async tasks that call the k2pow service, but not `await` on them.
- no parallel iteration is needed in this case, it can go through them sequentially
- depending on how many of those requests we need to do, we could just spawn all the futures all at once, and the k2pow service will make sure to execute (, throttle) them and return the results. on the post-service side, we could just use something like [this](https://docs.rs/futures/latest/futures/future/fn.try_join_all.html) to get the results, ordered.
- the `cores` param turns into, instead, a parameter that would be throttling the number of open requests to remote k2pow service/LB. we could use a `tokio::sync::Semaphore` to do that. The k2pow client (in `src/pow/service.rs`) would in turn use this parameter, and will therefore limit the amount of open in-flight requests (it would be rather a `parallelism` setting for api calls).
- then zip the iterator of the (ordered) results with the iterator of the inputs, generating the same vector of `AesCipher` we'd normally get from the sync execution context when performing the calculation locally.

Now... speaking of weaving this into what we already have, I suppose is a bit more complex. It also means plumbing in `async` way deeper into the call-stack into areas where either synchronous-blocking execution may also happen and I'm not sure how I feel about it, because right now the async context is completely missing from those areas in the codebase which, I find, makes it easier to reason about it better.

Maybe just having a `parallelism` setting that will just spawn a constant size rayon threadpool on which to execute all the api calls on would be easier, but at a cost of a lot of unnecessary overhead of thread context-switching.

Reply:
>The thing is that we don't need a thread pool here. Work stealing with little threads would do the same thing with far less context switching.

Rayon [uses work stealing](https://github.com/rayon-rs/rayon/blob/main/FAQ.md#how-does-rayon-balance-work-between-threads) :)

Regarding async, I don't feel this is a good idea to make `generate_proof` async. But maybe it would make sense to add another _blocking_ method to the `pow::Prover` trait that would calculate proofs of work for a given range of input nonces? Inside it could asynchronously spawn requests and await them. Btw, this is how blocking requests in `reqwest` work too (they expose a _sync_ API for the _async_ inner functions). See also a helpful post here: https://stackoverflow.com/questions/51044467/how-can-i-perform-parallel-asynchronous-http-get-requests-with-reqwest

>the cores param turns into, instead, a parameter that would be throttling the number of open requests to remote k2pow service.

I'd rather not repurpose existing configuration parameter doing other things for this. I think it would be best to have another configuration option for this with some sane default.

Reply:
> Rayon [uses work stealing](https://github.com/rayon-rs/rayon/blob/main/FAQ.md#how-does-rayon-balance-work-between-threads) :)

It does, in order to do a lot of _CPU-bound_ work (my understanding, also from reading around about rayon, is that it is meant for CPU bound use-cases, not IO bound). When we ask for a remote k2pow service to produce the results we need, there's zero computation involved, only waiting on the response of the http response. So, if you have, for example, 200 remote k2pow hosts to query from, you'll need 200 threads to do that in parallel, while in async context you'll need maybe a couple. That sounds pretty wasteful.

> Regarding async......

So here I'm not sure I'm fully onboard and this smells a bit, because:
- the main application is spawned as a tokio async application
- the tonic part is async as well
- the actual worker part is currently spawned as a blocking task within tokio (which is fine because currently it is 100% CPU bound)

The changes you're suggesting would mean (in the case of the k2pow service case):
- the main application is spawned as a tokio async application
- the tonic part is async as well
- worker part is currently spawned as a blocking task within tokio
- within the blocking task, a large threadpool is created
- within each thread, a blocking task is created
- the blocking task uses a blocking API that uses async under-the-hood

So to sum it up, I'm not 100% fan of spawning an async context in order to call a blocking context that calls an async context.

> I'd rather not repurpose existing configuration parameter doing other things for this. I think it would be best to have another configuration option for this with some sane default.

I meant that, sorry, my phrasing was somewhat misleading. I meant to use a differently named parameter.

Reply:
I think we agree that rayon is not the best match here :)

About
>The changes you're suggesting would mean (in the case of the k2pow service case):
> - the main application is spawned as a tokio async application
> - the tonic part is async as well
> - worker part is currently spawned as a blocking task within tokio
> - within the blocking task, a large threadpool is created
> - within each thread, a blocking task is created
> - the blocking task uses a blocking API that uses async under-the-hood

I think you misunderstood me. I meant something like this:
...
- worker part is currently spawned as a blocking task within tokio
- within the blocking task (`generate_proof`), call blocking pow::Prover::ProveMany, spawn async task per nonce_group and join them
- the blocking task uses a blocking API that uses async under-the-hood (using something like https://docs.rs/futures/latest/futures/executor/fn.block_on.html to run the above future to completion on the current thread).

There is no additional thread pool involved. By the way, this also happens under the hood in the `reqwest` blocking API anyways.